src/main/java/com/thealgorithms/strings/AhoCorasick.java

```diff
@@ -0,0 +1,227 @@
+                    if (parentSuffix.getChild().containsKey(cc)) {
+                        currentChild.setSuffixLink(parentSuffix.getChild().get(cc));
+                    } else {
+                        currentChild.setSuffixLink(root);
+                    }
+
+                    q.add(currentChild); // Add the child node to the queue for further processing
+                }
+
+                // Establish output links for nodes to efficiently identify patterns within patterns
+                if (currentState.getSuffixLink().getPatternInd() >= 0) {
+                    currentState.setOutputLink(currentState.getSuffixLink());
+                } else {
+                    currentState.setOutputLink(currentState.getSuffixLink().getOutputLink());
+                }
+            }
+        }
+
+        // Searches for patterns in the input text and records their positions
+        public ArrayList<ArrayList<Integer>> searchIn(String text) {
+            /*
+             * positionByStringIndexValue is the list of list containing the indexes of words in patterns/dictionary
+             * list index represents word.
+             * eg - list[0] contains the list of start-index of word pattern[0]
+             *      list[1] contains the list of start-index of word pattern[1]
+             *      ......
+             *      ......
+             *      list[n] contains the list of start-index of word pattern[n]
+             */
+            ArrayList<ArrayList<Integer>> positionByStringIndexValue = new ArrayList<>(patterns.length); // Stores positions where patterns are found in the text
+            Node parent = root; // Start searching from the root node
+
+            for (int i = 0; i < patterns.length; i++) {
+                positionByStringIndexValue.add(new ArrayList<Integer>()); // Initialize a list to store positions of the current pattern
+            }
+
+            for (int i = 0; i < text.length(); i++) {
+                char ch = text.charAt(i); // Get the current character in the text
+
+                // Check if the current node has a child for the current character
+                if (parent.getChild().containsKey(ch)) {
+                    parent = parent.getChild().get(ch); // Update the current node to the child node
+
+                    // If the current node represents a pattern, record its position in positionByStringIndexValue
+                    if (parent.getPatternInd() > -1) {
+                        positionByStringIndexValue.get(parent.getPatternInd()).add(i);
+                    }
+
+                    Node outputLink = parent.getOutputLink();
+                    // Follow output links to find and record positions of other patterns
+                    while (outputLink != null) {
+                        positionByStringIndexValue.get(outputLink.getPatternInd()).add(i);
+                        outputLink = outputLink.getOutputLink();
+                    }

```

Comment:
This ask to be in a separate method. In order to make it _nice_ I would suggest to create another class like `RawSearchResult`. It could be either derived from `ArrayList<ArrayList<Integer>>` (I would call it a _tricky_ approach) or simply a class having one member of the type `ArrayList<ArrayList<Integer>>`.

This comment is not a _must_.

Reply:
Hey @vil02 ,
I tried implementing what you said above. I used this
 
> simply a class having one member of the type `ArrayList<ArrayList<Integer>>`.

I hope you like it.
 
Still if there's any thing that needs to be added or if you want me to revert it back to how it was, please do tell me. I will address it asap.